pkg/util/otel/otel.go

```diff
@@ -0,0 +1,110 @@
+package otel
+
+import (
+	"context"
+
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects"
+	errors2 "github.com/pkg/errors"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric/noop"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type shutdownFn func(ctx context.Context) error
+
+// Start sets up and starts all components needed for creating OpenTelemetry traces and metrics as well as some common auto-instrumentation
+// It logs and swallows all errors to not prevent the application from startup.
+func Start(ctx context.Context, otelServiceName string, apiReader client.Reader, webhookNamespace string) func() {
+	endpoint, apiToken, err := getOtelConfig(apiReader, webhookNamespace)
+
+	if err != nil {
+		log.Error(err, "failed to read OTel config secret")
+		return setupNoopOTel()
+	}
+
+	shutdown, err := setupOtlpOTel(ctx, otelServiceName, endpoint, apiToken)
+	if err != nil {
+		log.Error(err, "failed to setup OTLP OTel")
+		return setupNoopOTel()
+	}
+	return shutdown
+}
+
+func setupOtlpOTel(ctx context.Context, otelServiceName string, endpoint string, apiToken string) (func(), error) {
+	otelResource, err := newResource(otelServiceName)
+	if err != nil {
+		return nil, err
+	}
+
+	_, tracesShutdownFn, err := setupTracesWithOtlp(ctx, otelResource, endpoint, apiToken)
+	if err != nil {
+		return nil, err
+	}
+
+	meterProvider, metricsShutdownFn, err := setupMetricsWithOtlp(ctx, otelResource, endpoint, apiToken)
+	if err != nil {
+		_ = tracesShutdownFn(ctx)
+		return nil, err
+	}
+	startAutoInstrumentation(meterProvider)
+
+	return func() {
+		_ = tracesShutdownFn(ctx)
+		_ = metricsShutdownFn(ctx)
+	}, nil
+}
+
+// setupNoopOTel makes sure, that OTel is properly configured so that no subsequent usage of OTel leads to panics while keeping
+// a minimal impact on runtime. Basically all collected metrics and traces get discarded right away.
+func setupNoopOTel() func() {
+	otel.SetMeterProvider(noop.NewMeterProvider())
+	otel.SetTracerProvider(trace.NewNoopTracerProvider())
+	log.Info("use Noop providers for OpenTelemetry")
+	return func() {}
+}
+
+// newResource returns a resource describing this application.
+func newResource(otelServiceName string) (*resource.Resource, error) {
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(otelServiceName),
+		),
+	)
+	return r, errors2.WithStack(err)
+}
+
+func getOtelConfig(apiReader client.Reader, namespace string) (string, string, error) {
```

Comment:
pass in the `ctx` pls (you should be using it in line 92)

Reply:
Fixed.

Reply:
done